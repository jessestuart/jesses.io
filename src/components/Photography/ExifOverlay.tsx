import React from 'react'
import { Flex, TextProps } from 'rebass/styled-components'
import styled from 'styled-components'
import { ExifData } from 'gatsby-source-s3-image'
import _ from 'lodash'

import { mapLensModelExif } from 'utils/exif'

const StyledExifOverlay = styled(Flex)<TextProps & { isActive: boolean }>`
  background: linear-gradient(
    0,
    rgba(0, 0, 0, 1) 0%,
    rgba(200, 200, 200, 0) 100%
  );
  bottom: 0;
  color: white;
  display: flex;
  font-size: 1rem;
  justify-content: flex-end;
  padding: 10px;
  position: absolute;
  text-align: right;
  text-shadow: 0 0 4px #000;
  transition: opacity 0.5s;
  width: 100%;
`

interface Props {
  exif: ExifData | undefined
  isActive: boolean
}

const getExifDescription = _.memoize((exif: ExifData) => {
  const lensModel: string | undefined = mapLensModelExif(exif.LensModel || '')
  const { FNumber, FocalLength, ISO, ShutterSpeedFraction } = exif
  return (
    <>
      {FocalLength ? `${FocalLength}mm, ` : null}
      {ShutterSpeedFraction ? `${ShutterSpeedFraction}s, ` : null}
      {FNumber ? `Æ’${FNumber}, ` : null}
      {ISO ? `ISO ${ISO}` : null}
      <br />
      {lensModel}
    </>
  )
})

const ExifOverlay = ({ exif, isActive }: Props) => {
  if (!exif) {
    return null
  }
  return (
    <StyledExifOverlay
      fontFamily="smallcaps"
      isActive={isActive}
      style={{ opacity: isActive ? 1 : 0 }}
    >
      {getExifDescription(exif)}
    </StyledExifOverlay>
  )
}

export default ExifOverlay
