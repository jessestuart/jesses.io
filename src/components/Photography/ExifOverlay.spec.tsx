import React from 'react'
import renderer from 'react-test-renderer'

import ExifOverlay from 'components/Photography/ExifOverlay'

test('Render valid inactive ExifOverlay component.', () => {
  const component = (
    <ExifOverlay isActive={false}>Shot with a camera</ExifOverlay>
  )
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Render valid active ExifOverlay component.', () => {
  const component = (
    <ExifOverlay isActive={true}>Shot with a camera</ExifOverlay>
  )
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Rendering ExifOverlay w/out any text is empty div.', () => {
  // @ts-ignore
  const tree = renderer.create(<ExifOverlay />).toJSON()
  expect(tree).toMatchSnapshot()
})
