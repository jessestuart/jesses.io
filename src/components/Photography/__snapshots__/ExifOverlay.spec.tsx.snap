// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render valid active ExifOverlay component. 1`] = `
<div
  className="sc-bdVaJa hChdkv"
>
  Shot with a camera
</div>
`;

exports[`Render valid inactive ExifOverlay component. 1`] = `
<div
  className="sc-bdVaJa hQVqai"
>
  Shot with a camera
</div>
`;

exports[`Rendering ExifOverlay w/out any text is empty div. 1`] = `
<div
  className="sc-bdVaJa hQVqai"
/>
`;
